### PR TITLE
Chapter 5: Python Code

### PR DESCRIPTION
**Pull request by Smit-create**

### CODE REVIEW

**Comment by Smit-create** [comment]
Results from firm exit:
firm_exit_1
firm_exit_2

File: source_code_py/firm_exit.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        β=0.98, s=100.0         # discount factor and scrap value
+    ):
+    """
+    Creates an instance of the firm exit model.
+    """
+    mc = tauchen(ρ, ν, b=μ, n=n)
+    z_vals, Q = mc.state_values, mc.P
+    return Model(n=n, z_vals=z_vals, Q=Q, β=β, s=s)
+
+
```
Comment (jstac) [inline source_code_py/firm_exit.py:28]: jit compile?

File: source_code_py/firm_exit.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+def no_exit_value(model):
+    """Compute value of firm without exit option."""
+    n, z_vals, Q, β, s = model
+    I = np.identity(n)
+    return np.linalg.solve((I - β * Q), z_vals)
+
+
```
Comment (jstac) [inline source_code_py/firm_exit.py:35]: jit compile?

File: source_code_py/firm_exit.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+def T(v, model):
+    """The Bellman operator Tv = max{s, π + β Q v}."""
+    n, z_vals, Q, β, s = model
+    h = z_vals + β * np.dot(Q, v)
+    return np.maximum(s, h)
+
+
```
Comment (jstac) [inline source_code_py/firm_exit.py:42]: jit compile?

**Review by jstac — APPROVED** [review]
Many thanks @Smit-create , great work.

Perhaps it's worth jit compiling some of the functions indicated below, just so the speed is comparable with the Julia code?  (I realize that these problems are small and the code executes quickly anyway.)

**Comment by Smit-create** [comment]
I modified the code slightly ( fixed some minor bugs) to make it jit compile and got the following results:
american_option_1
american_option_2

**Comment by Smit-create** [comment]
@jstac Does this look good now?

File: source_code_py/american_option.py [hunk 4]
```diff
@@ -7,12 +7,26 @@
...
 
 
 # NamedTuple Model
 Model = namedtuple("Model", ("t_vals", "z_vals","w_vals", "Q",
-                             "φ", "T", "β", "K", "e"))
+                             "φ", "T", "β", "K"))
+
+
+@njit
+def _eval(t, i_w, i_z, T, K):
```
Comment (jstac) [inline source_code_py/american_option.py:19]: @Smit-create , can you please call this function `exit_reward` and put in a docstring that says "Payoff to exercising the option at time t."

File: source_code_py/american_option.py [hunk 5]
```diff
@@ -7,12 +7,26 @@
...
+    return (t < T) * (i_z + i_w - K)
+
+
+@njit
+def product(a, b):
```
Comment (jstac) [inline source_code_py/american_option.py:24]: Can this function be replaced by something in https://quanteconpy.readthedocs.io/en/latest/tools/gridtools.html, such as `cartesian`?

**Comment by jstac** [comment]
Thanks @Smit-create , please see my comments above.

File: source_code_py/american_option.py (hunk 5 above)
Comment (Smit-create) [reply source_code_py/american_option.py:24]: Yes, that's the same, but using that will throw jit errors as:
```
Untyped global name 'cartesian': Cannot determine Numba type of <class 'function'>

File "american_option.py", line 56:
def C(h, model):
    <source elided>
    z_idx, w_idx = np.arange(len(z_vals)), np.arange(len(w_vals))
    for (t, i_z) in cartesian(t_vals, z_idx):
```

**Comment by jstac** [comment]
Thanks @Smit-create .  I wonder if we should modify `cartesian` to be jit-aware --- otherwise we will need to repeat your `product` function in quite a few files for this project.  It's just a suggestion and I'm open to other ideas.

Regarding the figures, I notice that the font size for "in the money" and "out of the money" is too large in the Python code (please see above).

Finally, for successive approximation, perhaps we should use https://quanteconpy.readthedocs.io/en/latest/tools/compute_fp.html

**Comment by Smit-create** [comment]
> Finally, for successive approximation, perhaps we should use https://quanteconpy.readthedocs.io/en/latest/tools/compute_fp.html

I was trying to avoid using direct library functions so that if anyone tries to compare python and Julia code, it would be easy to relate the things easily.

**Comment by Smit-create** [comment]
Fixed the plot
american_option_2

**Comment by Smit-create** [comment]
> I wonder if we should modify `cartesian` to be jit-aware

We can try to make that change in the `quantecon` itself? I mean we can try to make the functions/classes jit-aware as much as we can?

**Comment by Smit-create** [comment]
Thanks, @jstac. I have addressed your comments. If we replace `successive_approx` with `compute_fp` then we will need to do that in all the previous code to maintain uniformity and that will make it differ slightly from Julia code. Should I change that? Or keep the current version?

**Comment by jstac** [comment]
Thanks for the fixes @Smit-create 

> If we replace successive_approx with compute_fp then we will need to do that in all the previous code to maintain uniformity and that will make it differ slightly from Julia code. Should I change that? Or keep the current version?

I hadn't thought about this carefully enough.  How about this policy: we don't worry about matching Python and Julia code, as long as we reproduce the figures.

As a general rule, let's write the Python code as efficiently as possible, and with minimal repetition.  

Following this rule, let's use `compute_fp` here.  But there's no need to go back and change the earlier ones --- it might be good for people who are starting out to see all the details.

**Comment by Smit-create** [comment]
> As a general rule, let's write the Python code as efficiently as possible, with minimal repetition.

Thanks, @jstac. I just replaced the `product` and instead used loops with `prange` and also used `compute_fp` from `quantecon`. This works very fast now.

**Comment by jstac** [comment]
Very nice work @Smit-create .  I'll merge this now.

**Pull request merged**